docs/zh-CN/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,861 @@
+					onNameChange && onNameChange(newName);
+				}}
+			>
+				Set new name
+			</button>
+		</div>
+	);
+});
+```
+
+**基于类的部件：**
+
+```tsx
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+import { tsx } from '@dojo/framework/core/vdom';
+
+export interface MyWidgetProperties {
+	name: string;
+	onNameChange?(newName: string): void;
+}
+
+export default class MyWidget extends WidgetBase<MyWidgetProperties> {
+	private newName = '';
+	protected render() {
+		const { name, onNameChange } = this.properties;
+		return (
+			<div>
+				<span>Hello, {name}! Not you? Set your name:</span>
+				<input
+					type="text"
+					value={this.newName}
+					oninput={(e: Event) => {
+						this.newName = (e.target as HTMLInputElement).value;
+						this.invalidate();
+					}}
+				/>
+				<button
+					onclick={() => {
+						this.newName = '';
+						onNameChange && onNameChange(newName);
+					}}
+				>
+					Set new name
+				</button>
+			</div>
+		);
+	}
+}
+```
+
+此示例部件的使用者可以通过传入适当的属性与之交互：
+
+> src/widgets/NameHandler.tsx
+
+**基于函数的部件：**
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import icache from '@dojo/framework/core/middleware/icache';
+
+import MyWidget from './MyWidget';
+
+const factory = create({ icache });
+
+export default factory(function NameHandler({ middleware: { icache } }) {
+	let currentName = icache.get<string>('current-name') || 'Alice';
+	return (
+		<MyWidget
+			name={currentName}
+			onNameChange={(newName) => {
+				icache.set('current-name', newName);
+			}}
+		/>
+	);
+});
+```
+
+**基于类的部件：**
+
+```tsx
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+import { tsx } from '@dojo/framework/core/vdom';
+import watch from '@dojo/framework/core/decorators/watch';
+import MyWidget from './MyWidget';
+
+export default class NameHandler extends WidgetBase {
+	@watch() private currentName: string = 'Alice';
+
+	protected render() {
+		return (
+			<MyWidget
+				name={this.currentName}
+				onNameChange={(newName) => {
+					this.currentName = newName;
+				}}
+			/>
+		);
+	}
+}
+```
+
+## 高级：提取和注入状态
+
+实现复杂功能时，在部件内遵循状态封装模式可能会导致组件膨胀、难以管理。在大型应用程序中也可能出现另一个问题，数百个部件跨数十个层级组合在一起。通常是叶部件使用状态数据，并不是 VDOM 层次结构中的中间容器。让数据状态穿透这样一个层次结构复杂的部件需要增加脆弱、不必要的代码。
+
+Dojo 提供的 [Store 组件](/learn/stores/introduction) 解决了这些问题，它将状态管理提取到专用上下文中，然后将应用程序中的相关状态注入到特定的部件中。
+
+# 最佳开发实践
+
+使用 Dojo 部件时，应谨记一些重要原则，以避免在应用程序代码中引入反模式。试图以不受支持的方式使用框架可能会导致意外的行为，并在应用程序中引入难以发现的错误。
+
+## 部件属性
+
+-   部件应只能读取传入其中的属性(properties)。
+    -   如果修改了传入部件中的属性值，则不能回传给框架，以避免导致部件和框架之间出现差异。
+-   Widgets should avoid deriving further render state from their properties, and instead rely on their complete render state being provided to them.
+    -   Deriving render state can cause similar divergences between the widget and the framework as modifying received properties; the framework is not aware of the derived state, so cannot properly determine when a widget has been updated and requires invalidation and re-rendering.
```

Comment:
I do not understand this section.

Reply:
Hi @xiaohulu, thanks for these translations btw!

This section is essentially: don't expect Dojo to know about any derived state within a widget. Dojo only knows about the properties passed to a widget, so if any extra state is created & used for rendering, Dojo won't know when the extra state has been updated, so won't be able to re-render the widget. Widgets should be kept as simple as possible and rely on their state being passed to them.

Does that help?

Reply:
@sbinge Thanks. What does derived state means? Show some code that anti-pattern may help.

Reply:
Hi @xiaohulu - sorry for the delayed response, my github notifications have been playing up.

The following is a silly example of derived state:

```tsx

const factory = create().properties<{ currentValue: string }>();

function getDerivedState(value: string) {
	return `Current value is '${value}' and the current time is ${new Date().toISOString()}`;
}

export const WidgetWithDerivedState = factory(function WidgetWithDerivedState({ properties }) {
	const { currentValue } = properties();
	let derivedState = getDerivedState(currentValue);
	return (
		<div>
			<label for="dataEntry">Please enter a value:</label>
			<input
				id="dataEntry"
				value={currentValue}
				oninput={(e) => {
					derivedState = getDerivedState((e.target as HTMLInputElement).value);
				}}
			/>
			<div>{derivedState}</div>
		</div>
	);
});
```

There are two derived state anti-patterns here:

1. `getDerivedState()` uses data that changes independently of the widget's original state - in this example, it uses the current date/time.
1. the `oninput` event handler is changing the widget's state internally.

For both these anti-patterns, Dojo doesn't know that the widget's state has changed and therefore doesn't ever re-render the widget.

Reply:
@sbinge Thanks, it's so clear. I will translate this section.